File: R/monte_carlo.R
library(future.apply)

# Set up parallelization plan (assuming you want to use multisession for parallelism)
plan(multisession)

DIR <- "simulation_results"

simulate_and_save <- function(m, parameter_one, parameter_two) {
  set.seed(m)
  result <- parameter_one + parameter_two + rnorm(1, 0, 1)
  
  # Ensure the directory exists
  if(!dir.exists(DIR)) {
    dir.create(DIR, recursive = TRUE)
  }
  
  # Save results to csv
  write.csv(result, file = paste0(DIR, "/results_", m, ".csv"), row.names = FALSE)
  return(paste("Simulation", m, "saved."))
}

main <- function() {
  parameter_one <- 1
  parameter_two <- 2
  
  # Number of simulations
  num_simulations <- 64
  
  # Run simulations in parallel
  results <- future_lapply(1:num_simulations, function(m) simulate_and_save(m, parameter_one, parameter_two), future.seed = TRUE)
  
  print(results)
}

# Run the main function
main()

File: monte_carlo.py
from concurrent.futures import ProcessPoolExecutor
import os
import numpy as np

DIR = "simulation_results"

def simulate_and_save(m, paramter_one, paramter_two):
    # Simulate data and run model

    # Here, you would replace this with actual simulation code
    # For demonstration, we will just add two parameters and some random noise
    np.random.seed(m)
    result = paramter_one + paramter_two + np.random.normal(0, 1, 1)
    
    # Save results to csv
    np.savetxt(f"{DIR}/results_{m}.csv", result, delimiter=",")
    return f"Simulation {m} saved."

def main():
    paramter_one = 1
    paramter_two = 2
    
    # Number of simulations
    num_simulations = 64
    
    # Ensure the directory exists
    if not os.path.exists(DIR):
        os.makedirs(DIR, exist_ok=True)

    # Define the number of cores/processes to use
    num_cores = 4  # For example, to use 4 cores

    # Using ProcessPoolExecutor to run simulations in parallel
    with ProcessPoolExecutor(max_workers=num_cores) as executor:
        futures = [executor.submit(simulate_and_save, m, paramter_one, paramter_two) for m in range(num_simulations)]
        
        # Optionally, you can collect and print/log the return values from your simulations
        for future in futures:
            print(future.result())

if __name__ == "__main__":
    main()
